### PR TITLE
rgw/crypt: sse-s3 changes

### PR DESCRIPTION
**Pull request by mdw-at-linuxbox**
These are my sse-s3 changes, based on flipkart's work.  I tried to preserve flipkart's commits as much as possible, on the theory it might make review easier.  So 3 phases: what I had before, flipkart's work, and what I had after.

- phase 1 -
6683d27de2e rgw: sse-s3: Introducing SSE-S3 RGW options and using them
422df4a226e rgw/sse-s3: temporary code to test out sse-s3 vault operations - create and delete keys.
- phase 2 -
62499f14c27 rgw: Adding SSE-S3 support in GET and PUT paths (using Vault as KMS)
653a82e8836 rgw: Address review comments, fix test cases for rgw_kms
- phase 3 -
e429a38148b rgw/sse-s3: save sse attributes in req_state->crypt_attribute_map
56542900c8d rgw/sse-s3: sse-s3 should not require bucketencryption policy, work w/ postobj
031a7b04ab6 rgw/sse-s3: +get_encryption_defaults, use new crypt_attribute_map
cfcfa19812e rgw/sse-s3: +s3 get_encryption_defaults, postobj/crypt_attribute_map fixes
89c036e2323 rgw/sse-s3: bucketencryption should not generate bucket key names.
f3d8989d2e6 rgw/sse-s3: remove make_kek_s3/generate_kek_sse_s3
f462043ad30 rgw/sse-s3: make bucket key name configurable
1b6fdea414f rgw/sse-s3: +rgw_remove_sse_s3_bucket_key

In the above, commit 422df4a226e is intended to go away.  I'll make that go away when I address any review comments people might have.

### CODE REVIEW

File: src/rgw/rgw_op.cc
Comment (cbodley) [inline src/rgw/rgw_op.cc:3631]: i'm skeptical that bucket policy should see the transformed sse attributes (based on bucket encryption policy)

and it's not clear whether `get_encryption_defaults()` has performed those transformations yet

File: src/rgw/rgw_rest_s3.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_rest_s3.cc:2661]: good call, will fix.

File: src/rgw/rgw_kms.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_kms.cc:1266]: this is to mirror similar sse-kms code which allows multiple security engines - which I assume we will want here as well.

File: src/rgw/rgw_crypt.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_crypt.cc:940]: oops, good call.  will fix.

File: src/rgw/rgw_common.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_common.cc:435]: Yes.  I didn't think it was worth a string compare to do that.  I'd prefer to just use x_meta_map, but it seems to have a confusing set of uses and I didn't want to break that.

File: src/rgw/rgw_crypt.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_crypt.cc:1226]: I'm not sure what's right here, but of these I dislike EINVAL the least, so I'll use that.

File: src/common/options/rgw.yaml.in
Comment (mdw-at-linuxbox) [reply src/common/options/rgw.yaml.in:2770]: Problem is this shares code with kms, so I have to pass in a config name not a config value.  This is something that could be cleaned up in the future, perhaps as part of supporting additional security back ends (where multiple values might even make sense.)

File: src/rgw/rgw_crypt.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_crypt.cc:949]: Leaving it as '%' makes it less obvious something weird is happening, so I don't want to do that.  Argueably it should just fail, but that complicates the calling code and I'm not sure it actually provides better behavior.

File: src/rgw/rgw_kms.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_kms.cc:184]: um.  that is my code, and using templates instead could require a lot more runtime code, which has its own runtime costs.  Certainly worth more discussion.

File: src/rgw/rgw_op.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_op.cc:3631]: I believe get_encryption_defaults is called before this.  And it's unclear to me whether bucket policy sohuld see bucket encryption defaults or not.  I'm hopiong Ali will get us an ansewr on that.

File: src/rgw/rgw_common.cc
Comment (cbodley) [reply src/rgw/rgw_common.cc:435]: a string comparison here could avoid lots of small allocations from adding every header to this map. a potential alternative to this map would be a struct of std::strings for each of the known encryption headers, but that would add a lot of extra string comparisons here

regarding x_meta_map, i agree that we should preserve the original request headers there 👍

File: src/rgw/rgw_crypt.cc
Comment (cbodley) [reply src/rgw/rgw_crypt.cc:1226]: i'm surprised we don't have a mapping for `400 Bad Request` yet

File: src/rgw/rgw_kms.cc
Comment (cbodley) [reply src/rgw/rgw_kms.cc:184]: ah ok, i just recognized it from the original flipkart pr. let's defer this for after quincy

File: src/rgw/rgw_op.cc
Comment (cbodley) [reply src/rgw/rgw_op.cc:3631]: ok. in response to your email, i argued that the defaults shouldn't be visible to bucket policy. but that was based on an assumption that all request headers were visible to iam like `s3:x-amz-server-side-encryption`, but this function shows that's clearly not the case

i vote that we leave is this way for initial merge, and follow up with testing and docs based on what @alimaredia learns here

File: src/common/options/rgw.yaml.in
Comment (cbodley) [reply src/common/options/rgw.yaml.in:2770]: ok thanks, i see that `config_to_engine_and_parms()` takes this by string

File: src/rgw/rgw_crypt.cc
Comment (cbodley) [reply src/rgw/rgw_crypt.cc:949]: the user is unlikely to see the resulting string though. how about we leave the `%` as-is, but log a scary warning message about the missing substitution?

**Comment by mdw-at-linuxbox** [comment]
I've pushed an update that addresses many of the concerns above.  I pulled out the temporary hack to rgw-adm, so that's gone.  I've included minor fixes for documentation - there should be more, but the documentation on sse-kms seems to have been trimmed way down and is no more complete.  This version builds, I haven't done any other testing.

I've pulled in 2 commits from PR#44339.  Also note this includes a mutated version of what was in PR#42522 - that includes changes requested above here.

I've pushed a copy to ceph-ci as "mdw-mankey-10" so that should produce runable packages.

File: src/rgw/rgw_crypt.cc
Comment (cbodley) [inline src/rgw/rgw_crypt.cc:1474]: EINVAL

File: src/common/options/rgw.yaml.in
Comment (mdw-at-linuxbox) [reply src/common/options/rgw.yaml.in:2761]: making it a enum means removing the ability for these backends to take parameters, something I specifically want to allow, and put code in to make it possible.  there are options for vault key creation that I think we'll want to be able to set in the future.

File: src/rgw/rgw_crypt.cc
Comment (mattbenjamin) [inline src/rgw/rgw_crypt.cc:1230]: indent

File: src/rgw/rgw_crypt.cc
Comment (mattbenjamin) [inline src/rgw/rgw_crypt.cc:1477]: this is in the op-path, right?  is this something that needs heap allocation?

File: src/rgw/rgw_op.cc
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc:3420]: what benefit does the branch have, then?  should we cast the result to void or something?

File: src/rgw/rgw_op.cc
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc:3629]: ok; seems like this was a valuable cleanup

File: src/rgw/rgw_op.cc
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc:8621]: go lambda go

File: src/common/options/rgw.yaml.in
Comment (mattbenjamin) [reply src/common/options/rgw.yaml.in:2761]: seems ok to me?

File: src/rgw/rgw_crypt.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_crypt.cc]: doesn't matter, that's the original flipkart code (possibly slightly chewed up), it's replaced by a more s3 protocol compatible code path in a later commit.

File: src/rgw/rgw_crypt.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_crypt.cc]: possibly.  i left it in case it mattered to flipkart.

File: src/rgw/rgw_crypt.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_crypt.cc]: this gets moved and fixed in a later commit.

File: src/rgw/rgw_crypt.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_crypt.cc:887]: that's anti-bugging from flipkart, yes something like it should be there.

File: src/rgw/rgw_crypt.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_crypt.cc:891]: needs testing to see if this is right, but there's no reason to keep the comment, I'll make it go away.

File: src/rgw/rgw_crypt.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_crypt.cc:1230]: fixed in later commit.

File: src/rgw/rgw_crypt.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_crypt.cc:1477]: the same idiom is all over rgw_crypt.  if it's wrong it should be addressed all at once, probably separately from this PR.  Looks like there are some datatype visibility issues that would have to be resolved to combine storage allocations like you suggest.

File: src/rgw/rgw_op.cc
Comment (mdw-at-linuxbox) [reply src/rgw/rgw_op.cc:3420]: it documents that there is an error return, and that there's no sensible recovery at this point.  If the compiler cooperates, it may also present an opportunity to put a runtime breakpoint on it.  Conceivably future code could queue some sort of retry record.

**Comment by mdw-at-linuxbox** [comment]
I've pushed a new version.  I believe this one should address most of the issues people have raised above, including all the major ones.  It builds (for me), and was capable of using sse-s3 to store an object in my test environment.  It needs more testing of course.y

I've pushed a copy of this to ceph-ci, as "mdw-mankey-13".

**Comment by cbodley** [comment]
any updates here? are we passing the existing s3tests yet?

**Comment by mdw-at-linuxbox** [comment]
I've pushed a new version that has a companion set of s3 tests.  This version has various bug fixes for putbucketencryption including associated kms support.  The s3 tests are here: https://github.com/mdw-at-linuxbox/s3-tests/tree/sse-5

**Comment by mdw-at-linuxbox** [comment]
I've pushed a new version.  The most signficant change is to more thoroughly segregate rest parameters from bucketencryption configuration, and to properly detect missing or inconsistent rest parameters.

I have an corresponding adaption of s3 tests for this: repo https:github.com/mdw-at-linuxbox/s3-tests branch sse-s3-5 .  This can be merged to s3-tests after this set of changes is merged to ceph.

I've run this one through teuthology (testing against my copy of s3-tests):

just rgw:crypt (ali's settings):
http://pulpito.front.sepia.ceph.com/mwatts-2022-03-09_19:52:17-rgw:crypt-mdw-mankey-17-distro-default-smithi/
a more complete run:
http://pulpito.front.sepia.ceph.com/mwatts-2022-03-14_07:49:20-rgw-mdw-mankey-17-distro-default-smithi/
The complete run has 48 failures from "s3-tests-java" (some sorrt of gradle issue), and 5 other misc. failures, none of which I believe to be related to sse.

github is currently complaining that "make check" and "ceph api tests" is failing, but reports "no problems" for its reason why.  ?

**Comment by mattbenjamin** [comment]
jenkins test make check

**Comment by mattbenjamin** [comment]
jenkins test api

File: src/rgw/rgw_rest_s3.cc
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc:188]: nit: what's with the lack of indendation for all these douts? were they just for debugging, or did you intend for them to merge this way?

**Comment by cbodley** [comment]
looks great. thanks for adding the tests for BucketEncryption+KMS. tests are a bit lacking in the error cases, but otherwise the coverage looks good

**Comment by cbodley** [comment]
> @cbodley @mdw-at-linuxbox @alimaredia is this ready for merge, then?

the condition for merge is a clean test run, and there are still java test failures to resolve

**Comment by cbodley** [comment]
jenkins test api

**Comment by mdw-at-linuxbox** [comment]
I've pushed an update, mostly to see if this fixes some of the above github complaints which I believe are not sse-s3 specific.  This is just a rebase on a newer master.  Teuthology run here,
http://pulpito.front.sepia.ceph.com:80/mwatts-2022-04-08_00:09:12-rgw-mdw-mankey-19-distro-default-smithi/

3 teuthology problems I think - something with tempest that may not be sse-s3 related, problems in s3tests w/o vault - sse-s3 tests are run that probably shouldn't be, and problems with s3test-java - test is probably wrong.

**Comment by mdw-at-linuxbox** [comment]
I've made PRs for the s3 tests:
https://github.com/ceph/s3-tests/pull/440
https://github.com/ceph/java_s3tests/pull/11

**Comment by cbodley** [comment]
> 3 teuthology problems I think - something with tempest that may not be sse-s3 related

confirmed unrelated, i saw the same 4 tempest failures (container metadata and object expiry) in a recent batch of master prs https://pulpito.ceph.com/cbodley-2022-04-07_12:45:46-rgw-wip-cbodley-testing-distro-default-smithi/

File: qa/tasks/s3tests.py
Comment (cbodley) [inline qa/tasks/s3tests.py:436]: > problems in s3tests w/o vault - sse-s3 tests are run that probably shouldn't be

@mdw-at-linuxbox in this teuthology piece, we filter out the s3test cases tagged `sse-s3` when we're not configured for it. in s3tests, we can use the decorator `@attr('sse-s3')` to mark these. just make sure your new test cases have that too

**Comment by mdw-at-linuxbox** [comment]
I've pushed out a new version.  This version includes spiffy logic to detect when sse-s3 is available and enables the sse-s3 tests in s3tests if so.  For java I do not have a way to do the same thing, so i've pushed out a version of the java tests that does not run the one test that fails w/o sse-s3.  See my java s3 tests PR for more info on that.

**Comment by cbodley** [comment]
> This version includes spiffy logic to detect when sse-s3 is available and enables the sse-s3 tests in s3tests if so

in `qa/tasks/s3tests.py`, right? i'm not seeing here in the PR

**Comment by alimaredia** [comment]
@cbodley this branch seems to have the latest changes he's made to teuthology related things in this PR. https://github.com/mdw-at-linuxbox/ceph/tree/wip-master-mankey-22

Not sure why it's not reflected here yet, but I'm guessing it will soon.

File: qa/tasks/s3tests.py
Comment (alimaredia) [reply qa/tasks/s3tests.py:436]: @mdw-at-linuxbox did this comment get resolved?

**Review by alimaredia — APPROVED** [review]
https://pulpito.ceph.com/amaredia-2022-04-29_03:02:44-rgw-mdw-mankey-22-distro-basic-smithi/

**Pull request merged**
